### PR TITLE
serial: use flock instead of lock files

### PR DESCRIPTION
**Pull request by lynxeye-dev**
Some distributions like Fedora don't allow users access to /var/lock and
instead want applications to use lockdev, which in turn isn't available
in distributions like Debian, as those rightfully state [1] that there is a
much more convenient and and easier to get right interface available in
the form of flock(2).

Switch microcom to use flock instead of the hand rolled lockfiles, which
also avoids the issue of stale lock files, as the lock is automatically
removed when the application closes the serial fd.

[1] https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=734086

### CODE REVIEW

**Review by a3f — CHANGES_REQUESTED** [review]
I did a short test on my Debian 11 host:

- minicom 2.8: uses lock files
- picocom v3.1: use flock

Given that we can expect more tools to gradually use flock, I am in favor of this change.
The commit message should be explicit though, that we will no longer lock against some tools, like minicom.
The commit message and link read as if Debian already embraced the brave new world, but apparently there are still hold outs. (On a related note, can you run strace for minicom and find out what it does on Fedora?)

File: serial.c [hunk 1]
```diff
@@ -177,51 +161,20 @@ struct ios_ops * serial_init(char *device)
...
-		serial_unlock();
+	if (fd < 0)
 		main_usage(2, "cannot open device", device);
+
+	/* try to lock the device */
+	ret = flock(fd, LOCK_EX | LOCK_NB);
+	if (ret) {
+		if (!opt_force)
+			main_usage(3, "could not lock port", device);
+		else
```
[the remark this answers is not in the document]
Comment (lynxeye-dev) [reply serial.c:176]: I don't see what you want dropped here: either we stop execution via main_usage, or print a message and continue depending on the force option.

[the remark this answers is not in the document]
Comment (a3f) [reply serial.c:176]: The `else` itself can be dropped, not its content. Just a nitpick.

**Comment by ukleinek** [comment]
Alternative approach: https://github.com/ukleinek/microcom/tree/rework-locking

**Comment by lynxeye-dev** [comment]
> Alternative approach: https://github.com/ukleinek/microcom/tree/rework-locking

Sorry, but I just don't see the value of keeping the lock file stuff around. Even slow moving Distros like Debian want to get away from this way of locking serial devices (see the linked Debian Bug report). Please let us move microcom to thetechnology of the current millennium and drop the complexity introduced by lockfiles.

**Pull request merged**